### PR TITLE
DrivAerML: eta_min=5e-5 + gc=1.0 compound (stabilized non-zero LR floor)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Your previous experiment (PR #3126) showed that eta_min=5e-5 enables faster early convergence on DrivAerML (6.800% at ep98 vs champion at ~12% at same point), but diverged because removing the LR "rest" at cosine troughs allows gradient noise to accumulate unchecked.

This follow-up combines eta_min=5e-5 with gc=1.0 to provide the missing stability. The gc caps gradient norms at each step, preventing the accumulation that killed the eta_min-only run. The combination should preserve the faster early convergence while enabling the 400+ epoch training needed to reach baseline territory.

Note: gc alone on DM has been tested — best gc result was 4.346% (gc=2.0, PR #2886). eta_min alone diverged at 6.800%. The compound hypothesis is that gc provides the stability floor that eta_min removes, unlocking the faster convergence signal.

## Instructions

**Run 1: eta_min=5e-5 + gc=1.0**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --cosine-eta-min 5e-5 \
  --grad-clip 1.0 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group dm-eta-min-gc-compound
```

**Run 2: eta_min=5e-5 + gc=0.5 (lighter clipping)**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --cosine-eta-min 5e-5 \
  --grad-clip 0.5 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group dm-eta-min-gc-compound
```

If --cosine-eta-min was already added for #3126, reuse it. Otherwise add the flag and pass to CosineAnnealingLR's eta_min parameter. Report best val_primary/surface_rel_l2_pct.

## Baseline

- **val_primary/surface_rel_l2_pct:** 3.997% at epoch 467
- **Config:** 4L/512d/8H, AdamW lr=5e-4, T_max=30, eta_min=0, no gc, no-EMA
- **W&B run:** bht6h42t
- **Your #3126 result:** eta_min=5e-5 alone → 6.800% (faster early convergence, diverged)
- **Best gc result:** gc=2.0 → 4.346% (PR #2886)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT gohan:

## Results

### Run 1: eta_min=5e-5 + gc=1.0 (W&B: h6jblqjq)

| Metric | Value |
|--------|-------|
| **best val_primary/surface_rel_l2_pct** | **6.311%** (ep 149) |
| best test_primary/surface_rel_l2_pct | 7.354% (ep 149) |
| Baseline | 3.997% (ep 467) |
| Delta vs baseline | +57.8% (worse) |
| Total epochs | 501 (360-min timeout) |
| Peak GPU memory | ~6 GB |
| Diverged at | epoch 150 → 112.8% by ep 200 |

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --cosine-eta-min 5e-5 --grad-clip 1.0 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 --max-train-batches 394 \
  --max-eval-batches 200 --agent gohan \
  --wandb-name "gohan/dm-eta-min-5e5-gc-1.0" \
  --wandb-group dm-eta-min-gc-compound
```

**Trajectory (gc=1.0):**
- ep 50: 12.3% (gc_events: 73/394)
- ep 100: 7.4% (gc_events: 2/394)
- **ep 149: 6.311%** (gc_events: 1/394) ← best
- ep 150: 11.6% (gc_events: 62/394) ← **catastrophic divergence**
- ep 200+: ~112% (gc_events: 394/394, every batch clipped)

### Run 2: eta_min=5e-5 + gc=0.5 (W&B: 7fmjg0cu)

| Metric | Value |
|--------|-------|
| **best val_primary/surface_rel_l2_pct** | **8.617%** (ep 70) |
| best test_primary/surface_rel_l2_pct | 9.475% (ep 70) |
| Baseline | 3.997% (ep 467) |
| Delta vs baseline | +115.6% (worse) |
| Total epochs | 504 (360-min timeout) |
| Peak GPU memory | ~6 GB |
| Diverged at | epoch ~80 → 63.8% by ep 100 |

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --cosine-eta-min 5e-5 --grad-clip 0.5 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 --max-train-batches 394 \
  --max-eval-batches 200 --agent gohan \
  --wandb-name "gohan/dm-eta-min-5e5-gc-0.5" \
  --wandb-group dm-eta-min-gc-compound
```

**Trajectory (gc=0.5):**
- ep 50: 12.5% (gc_events: 288/394)
- **ep 70: 8.617%** (gc_events: 199/394) ← best
- ep 100: 63.8% (gc_events: 394/394) ← **diverged**
- ep 150+: ~72% (saturated at high loss plateau)

## What happened

**Both runs diverged — hypothesis FALSIFIED.**

Gradient clipping delayed but did not prevent the divergence that eta_min=5e-5 causes. The pattern is consistent across both clipping strengths:

1. **gc=1.0 delayed divergence to epoch 149** (vs gc=0.5 at epoch ~80, and the eta_min-only run from #3126 at ~98 epochs). It also reached a better pre-divergence minimum (6.311% vs 8.617% for gc=0.5 and 6.800% for eta_min-only).

2. **gc=0.5 was actively harmful** — tighter clipping constrained useful gradient signal during early training (288/394 batches clipped even at ep 50), leading to worse peak performance and earlier divergence.

3. **The divergence is catastrophic and unrecoverable** — once it begins, gradient norms explode (every batch clipped), loss jumps to a high plateau, and the model never recovers despite continued training for 300+ more epochs.

4. **Root cause analysis:** The non-zero LR floor at cosine troughs prevents the "rest" periods that allow the model to consolidate. Around cosine peaks (lr=5e-4), the high LR + residual momentum from the non-zero trough creates gradient instability. gc caps individual step norms but can't prevent the cumulative drift across many steps at the trough LR. The divergence consistently happens around epoch 150 (5 full cosine cycles) for gc=1.0, suggesting the instability compounds over multiple cycles.

5. **Comparison to gc-only results:** gc=2.0 alone (PR #2886) achieved 4.346% without divergence, confirming gc isn't the problem — eta_min is the destabilizing factor.

## Suggested follow-ups

- **Warmup restart at trough:** Instead of eta_min > 0, try a brief LR warmup at the start of each cosine cycle (warm restart / SGDR-style). This gives the "fresh start" momentum without the continuous noise at troughs.
- **Decaying eta_min:** Start with eta_min=5e-5 but anneal it toward 0 over training. This captures the early-convergence benefit while restoring the rest periods later.
- **gc=1.0 with eta_min=0 long training:** Given gc=1.0 showed the best gradient health (only 1 clip event at ep 149 before divergence), testing gc=1.0 on the standard eta_min=0 config might improve on the 3.997% baseline by providing stability in late training without the eta_min instability.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3132 because ADVISOR: eta_min=5e-5 is dead regardless of gc value. gc=1.0: 6.311% ep149 then diverged to 113% (W&B: h6jblqjq). gc=0.5: 8.617% ep70 then diverged to 72% (W&B: 7fmjg0cu). Third confirmation of this lineage — eta_min=5e-5 tested alone (#3126), +gc=1.0, and +gc=0.5, all diverge catastrophically. The non-zero LR floor prevents gradient momentum dissipation at cosine troughs. Adding to dead ends. Good root cause analysis on the gradient clip dynamics..
